components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/HybridPolicyDAOImpl.java

```diff
@@ -0,0 +1,253 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.entitlement.dao;
+
+import org.wso2.carbon.identity.entitlement.EntitlementException;
+import org.wso2.carbon.identity.entitlement.EntitlementUtil;
+import org.wso2.carbon.identity.entitlement.dto.AttributeDTO;
+import org.wso2.carbon.identity.entitlement.dto.PolicyDTO;
+import org.wso2.carbon.identity.entitlement.dto.PolicyStoreDTO;
+import org.wso2.carbon.identity.entitlement.policy.finder.AbstractPolicyFinderModule;
+import org.wso2.carbon.identity.entitlement.policy.finder.PolicyFinderModule;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * HybridPolicyDAOImpl is a hybrid implementation of PolicyDAO. It uses both JDBC and Registry implementations to handle
+ * policy data. If the policy is already in the registry, it will be maintained there, including new versions.
+ * New policies will be persisted in the database.
+ */
+public class HybridPolicyDAOImpl extends AbstractPolicyFinderModule implements PolicyDAO {
+
+    private final JDBCPolicyDAOImpl jdbcPolicyDAO = new JDBCPolicyDAOImpl();
+    private final RegistryPolicyDAOImpl registryPolicyDAO = new RegistryPolicyDAOImpl();
+
+    @Override
+    public void init(Properties properties) {
+
+        jdbcPolicyDAO.init(properties);
+        registryPolicyDAO.init(properties);
+    }
+
+    @Override
+    public void addOrUpdatePolicy(PolicyDTO policy, boolean isPapUpdate) throws EntitlementException {
+
+        if (registryPolicyDAO.isPolicyExistsInPap(policy.getPolicyId())) {
+            registryPolicyDAO.addOrUpdatePolicy(policy, isPapUpdate);
+        } else {
+            jdbcPolicyDAO.addOrUpdatePolicy(policy, isPapUpdate);
+        }
+    }
+
+    @Override
+    public PolicyDTO getPAPPolicy(String policyId) throws EntitlementException {
+
+        PolicyDTO policyDTO = jdbcPolicyDAO.getPAPPolicy(policyId);
+        if (policyDTO == null) {
+            policyDTO = registryPolicyDAO.getPAPPolicy(policyId);
+        }
+        return policyDTO;
+    }
+
+    @Override
+    public List<PolicyDTO> getPAPPolicies(List<String> policyIds) throws EntitlementException {
+
+        List<PolicyDTO> policyDTOs = jdbcPolicyDAO.getPAPPolicies(policyIds);
+        List<PolicyDTO> regPolicyDTOs = registryPolicyDAO.getPAPPolicies(policyIds);
+        return EntitlementUtil.mergeAndRemoveDuplicates(policyDTOs, regPolicyDTOs);
+    }
+
+    @Override
+    public PolicyDTO getPolicy(String policyId, String version) throws EntitlementException {
+
+        PolicyDTO policyDTO;
+        try {
+            policyDTO = jdbcPolicyDAO.getPolicy(policyId, version);
+        } catch (EntitlementException e) {
```

Comment:
Better if we can depend on returning  null `policyDTO` or check for isExists() rather than handling with exceptions.